### PR TITLE
PHOENIX-6413 Having cannot resolve alias

### PR DESCRIPTION
**Pull request by ZhaoBQ**

### CODE REVIEW

**Comment by ZhaoBQ** [comment]
I can not run test in local. I use command 'mvn clean package -Dtest=GroupByIT', and the error:
`[INFO] Running org.apache.phoenix.end2end.GroupByIT
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 42.827 s <<< FAILURE! - in org.apache.phoenix.end2end.GroupByIT
[ERROR] org.apache.phoenix.end2end.GroupByIT  Time elapsed: 42.818 s  <<< ERROR!
java.lang.RuntimeException: java.io.IOException: Shutting down
        at org.apache.phoenix.query.BaseTest.initMiniCluster(BaseTest.java:549)
        at org.apache.phoenix.query.BaseTest.setUpTestCluster(BaseTest.java:449)
        at org.apache.phoenix.query.BaseTest.checkClusterInitialized(BaseTest.java:435)
        at org.apache.phoenix.query.BaseTest.setUpTestDriver(BaseTest.java:517)
        at org.apache.phoenix.query.BaseTest.setUpTestDriver(BaseTest.java:512)
        at org.apache.phoenix.end2end.ParallelStatsDisabledIT.doSetup(ParallelStatsDisabledIT.java:62)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.apache.phoenix.SystemExitRule$1.evaluate(SystemExitRule.java:40)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.io.IOException: Shutting down
        at org.apache.hadoop.hbase.MiniHBaseCluster.init(MiniHBaseCluster.java:266)
        at org.apache.hadoop.hbase.MiniHBaseCluster.<init>(MiniHBaseCluster.java:115)
        at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniHBaseCluster(HBaseTestingUtility.java:1177)
        at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:1140)
        at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:1094)
        at org.apache.phoenix.query.BaseTest.initMiniCluster(BaseTest.java:544)
        ... 28 more
Caused by: java.lang.RuntimeException: Master not active after 30000ms
        at org.apache.hadoop.hbase.util.JVMClusterUtil.waitForEvent(JVMClusterUtil.java:232)
        at org.apache.hadoop.hbase.util.JVMClusterUtil.startup(JVMClusterUtil.java:188)
        at org.apache.hadoop.hbase.LocalHBaseCluster.startup(LocalHBaseCluster.java:430)
        at org.apache.hadoop.hbase.MiniHBaseCluster.init(MiniHBaseCluster.java:259)
        ... 33 more
`

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   5m 13s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  16m 38s |  master passed  |
| +0 |  hbaserecompile  |  24m 19s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  2s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 38s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 52s |  master passed  |
| +0 :ok: |  spotbugs  |   3m 12s |  phoenix-core in master has 959 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   8m 40s |  the patch passed  |
| +0 |  hbaserecompile  |  18m 37s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  1s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  1s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 37s |  phoenix-core: The patch generated 3 new + 265 unchanged - 3 fixed = 268 total (was 268)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 23s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 109m 27s |  phoenix-core in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 45s |  The patch does not generate ASF License warnings.  |
|  |   | 169m 56s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | phoenix.end2end.ConcurrentUpsertsWithoutIndexedColsIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1168 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaserebuild hbaseanti checkstyle compile |
| uname | Linux ef8a1b7996ab 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / ebe46e1 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/1/artifact/yetus-general-check/output/patch-unit-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/1/testReport/ |
| Max. process+thread count | 11849 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   6m 29s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  17m 20s |  master passed  |
| +0 |  hbaserecompile  |  24m 18s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  5s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 37s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 53s |  master passed  |
| +0 :ok: |  spotbugs  |   3m 17s |  phoenix-core in master has 959 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   9m 54s |  the patch passed  |
| +0 |  hbaserecompile  |  20m 11s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  5s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  5s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 35s |  phoenix-core: The patch generated 3 new + 265 unchanged - 3 fixed = 268 total (was 268)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 54s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 29s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 112m 53s |  phoenix-core in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   | 176m  1s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | phoenix.end2end.UpsertSelectIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1168 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaserebuild hbaseanti checkstyle compile |
| uname | Linux f153e9eeba19 4.15.0-136-generic #140-Ubuntu SMP Thu Jan 28 05:20:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / ebe46e1 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/2/artifact/yetus-general-check/output/patch-unit-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/2/testReport/ |
| Max. process+thread count | 11199 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1168/2/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-core/src/main/java/org/apache/phoenix/parse/ParseNodeRewriter.java [hunk 1]
```diff
@@ -157,6 +152,14 @@ public static SelectStatement rewrite(SelectStatement statement, ParseNodeRewrit
             }
             normGroupByNodes.add(normGroupByNode);
         }
+
+        ParseNode having = statement.getHaving();
```
Comment (lhofhansl) [inline phoenix-core/src/main/java/org/apache/phoenix/parse/ParseNodeRewriter.java:156]: Do this is just an ordering problem? Looks good to me.

**Review by lhofhansl — APPROVED** [review]

File: phoenix-core/src/main/java/org/apache/phoenix/parse/ParseNodeRewriter.java (hunk 1 above)
Comment (ZhaoBQ) [reply phoenix-core/src/main/java/org/apache/phoenix/parse/ParseNodeRewriter.java:156]: Thanks @lhofhansl . Yes, just an ordering problem. After adjusting the order, we can resolve alias in having clause.

**Review by gjacoby126 — APPROVED** [review]
+1, thanks for this patch, @ZhaoBQ

**Pull request merged**